feature/4743_markdown_support_flow_diagrams

jonnymoo opened the pull request:
## :bookmark_tabs: Summary

Brief description about the content of your PR.

Resolves #<your issue id here>

## :straight_ruler: Design Decisions

Describe the way your implementation works or what design decisions you made if applicable.

### :clipboard: Tasks

Make sure you

- [ ] :book: have read the [contribution guidelines](https://github.com/mermaid-js/mermaid/blob/develop/CONTRIBUTING.md)
- [ ] :computer: have added necessary unit/e2e tests.
- [ ] :notebook: have added documentation. Make sure [`MERMAID_RELEASE_VERSION`](https://github.com/mermaid-js/mermaid/blob/develop/packages/mermaid/src/docs/community/development.md#3-update-documentation) is used for all new features.
- [ ] :bookmark: targeted `develop` branch

jonnymoo commented on vite.config.ts:
```diff
@@ -16,8 +16,6 @@ export default defineConfig({
   test: {
     environment: 'jsdom',
     globals: true,
-    // TODO: should we move this to a mermaid-core package?
-    setupFiles: ['packages/mermaid/src/tests/setup.ts'],
```
I don't think so - it was in there to do a global set up for the d3 mock which is no longer required.

nirname replied:
Please, do not resolve conversation by yourself, this makes it more difficult to find what question has been addressed and which one has not

jonnymoo commented on packages/mermaid/src/diagrams/timeline/svgDraw.js:
```diff
@@ -523,23 +524,40 @@ export const drawNode = function (elem, node, fullSection, conf) {
 
   // Create the wrapped text element
   const textElem = nodeElem.append('g');
+  const description = node.descr.replace(/(<br\/*>)/g, '\n');
```
Good call - let me put a couple of tests round it and I'll recommit.

nirname commented on docs/ecosystem/integrations-community.md:
```diff
@@ -39,6 +39,8 @@ Below are a list of community plugins and integrations created with Mermaid.
   - [Mermaid plugin for GitBook](https://github.com/wwformat/gitbook-plugin-mermaid-pdf)
 - [LiveBook](https://livebook.dev) ✅
 - [Atlassian Products](https://www.atlassian.com)
+  - [Mermaid Charts & Diagrams for Confluence](https://marketplace.atlassian.com/apps/1222572/)
+  - [Mermaid Charts & Diagrams for Jira](https://marketplace.atlassian.com/apps/1224537/)
```
Nothing against it, but this is a bit out of the scope for the task. Let it be here, since you've already added it, but it's better to keep PR directly related to the issue, that would make tracking  the issue down much easier

jonnymoo replied:
Oh - yeh - I agree. I'm not sure why it is here either? I haven't added it on purpose - has it come through on a merge? More than happy for them to be removed.

nirname replied:
It is probably originates from the merge commit, yep

sidharthv96 replied:
There was a commit which didn't update the docs. Most probably this was added by the pre-commit hook. 
You can ignore these. It will go away once you sync the latest develop.

nirname reviewed (approved):
Let's wait till someone else will approve it, always better to verify it twice.

Yokozuna59 commented on packages/mermaid/src/diagrams/timeline/timeline.spec.js:
I wouldn't recommend testing this feature like this, you're kinda testing `createText` right now.

jonnymoo replied:
This is a good point - the actual thing I am putting under test is svgDraw.drawNode as this was the smallest self contained unit that was easy to test. If this not the appropriate domain level thing to test behaviour at, there are a couple of options.

The rest of the (previously existing) tests in timeline.spec.js are testing the parser. The tests won't fit here as they are to do with rendering not semantics.

I could test a level higher and test draw in timelineRenderer.ts (although this becomes tricky as a self enclosed unit because it goes and fetches config). Or...

The other option is to not unit test and test the behaviour only with e2e tests. If you worried that I am too close to testing implementation this might be appropriate?

Yokozuna59 reviewed (commented):
Can you please update the timeline docs showing that it's possible to do such thing? You can take inspiration from [flowchart](https://mermaid.js.org/syntax/flowchart.html) and [mindmap](https://mermaid.js.org/syntax/mindmap.html) docs.

And can you please verify that other string types of markdown are supported? i.e., italic.
If they're supported, then please add them to [`cypress/integration/rendering/timeline.spec.ts`](https://github.com/mermaid-js/mermaid/blob/775ab1b18ecea7e2b4befb2af193d7a0defca98d/cypress/integration/rendering/timeline.spec.ts)

nirname commented on packages/mermaid/src/diagrams/timeline/svgDraw.js:
```diff
@@ -523,23 +524,40 @@ export const drawNode = function (elem, node, fullSection, conf) {
+  if (!htmlLabels) {
+    const dx = node.width / 2;
+    const dy = node.padding / 2;
+    textElem.attr('transform', 'translate(' + dx + ', ' + dy + ')');
+  } else {
+    const dx = (node.width - bbox.width) / 2;
+    const dy = (node.height - bbox.height) / 2;
+    textElem.attr('transform', 'translate(' + dx + ', ' + dy + ')');
+  }
```
No, it cannot. Scope of the variables is inside `if` statement only. They must be not `const` but `var` then and be declared before `if` statement.

Yokozuna59 replied:
I see, I thought that if we have all possible cases then it would recognize it as an already declared variable. In this case I'm not sure if we should modify it to this:

```ts
let dx: number, dy: number;
if (!htmlLabels) {
   dx = node.width / 2;
   dy = node.padding / 2;
 } else {
   dx = (node.width - bbox.width) / 2;
   dy = (node.height - bbox.height) / 2;
 }
textElem.attr('transform', `translate(${dx}, ${dy})`);
```

sidharthv96 replied:
The number of lines is same in both cases, and there are two more variables in the outer scope. 
So I don't think the current implementation should definitely be changed. 


If making things compact is a goal, we could do something like this. 
```ts
   const dx = (node.width - (htmlLabels ? bbox.width : 0)) / 2;
   const dy = (node.height - (htmlLabels ? bbox.height : 0)) / 2;
   textElem.attr('transform', `translate(${dx}, ${dy})`);
```

jonnymoo replied:
I'll tidy this up - I'm can't remember why we don't include the bounding box for svgs - it may have just been attempting to make it consistent with before. I'll work this out first and then either simplify or use as per @sidharthv96 suggested code.

jonnymoo replied:
And I'll add the documentation / additional tests as per top of conversation.

sidharthv96 commented on packages/mermaid/src/diagrams/timeline/styles.js:
```diff
@@ -77,5 +77,17 @@ const getStyles = (options) =>
   .eventWrapper  {
    filter: brightness(120%);
   }
+  .timeline-node-label {
+    dy: 1em;
+    alignment-baseline: middle;
+    text-anchor: middle;
+    dominant-baseline: middle;
+    text-align: center;
+  }
```
I'd prefer the stylings to be in the styles file, instead of inside the draw.

jonnymoo commented on packages/mermaid/src/diagrams/timeline/svgDraw.js:
```diff
@@ -523,23 +524,40 @@ export const drawNode = function (elem, node, fullSection, conf) {
-    .attr('alignment-baseline', 'middle')
-    .attr('dominant-baseline', 'middle')
-    .attr('text-anchor', 'middle')
-    .call(wrap, node.width);
```
It's only required of htmlLabels is false, other the styling is done in css and is in the style.js file. (See the conversation below).

nirname commented:
@jonnymoo I'd like to revive this PR, could you update your branch to the latest development branch?

jonnymoo commented:
Hey Nikolay - yes will do. Apologies, I've been onto the project for a little while, so I'll just need to get my head back into it. I'll try and get this done in the next week.

Cheers,

Jonny
________________________________
From: Nikolay Rozhkov ***@***.***>
Sent: 20 June 2024 15:20
To: mermaid-js/mermaid ***@***.***>
Cc: Jonny Muir ***@***.***>; Mention ***@***.***>
Subject: Re: [mermaid-js/mermaid] feature/4743_markdown_support_timeline_diagrams (PR #4852)


@jonnymoo<https://github.com/jonnymoo> I'd like to revive this PR, could you update your branch to the latest development branch?

—
Reply to this email directly, view it on GitHub<https://github.com/mermaid-js/mermaid/pull/4852#issuecomment-2180835508>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AASGWJM4Q3GJJHZVMQX6RH3ZILQKNAVCNFSM6AAAAAA4446MT2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCOBQHAZTKNJQHA>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

jonnymoo commented:
I have rebased and cleaned up some merge conflict. I'm unsure if a rebase was the correct thing to do as this pull request now looks like it for 241 files - which isn't correct. Let me know if there is anything else I need to do.

jeswr commented:
I would find utility in this work being picked back up, and also for the ability to have hyperlinks within SVG digrams when doing so.
